File: presentations/14-repository-and-builder-pattern-demos/java-smart-home-scenes/src/main/java/edu/ksu/swe4743/smarthomescenes/builder/DeviceSceneBuilder.java
package edu.ksu.swe4743.smarthomescenes.builder;

import edu.ksu.swe4743.smarthomescenes.domain.DeviceGroupTarget;
import edu.ksu.swe4743.smarthomescenes.domain.DeviceScene;
import edu.ksu.swe4743.smarthomescenes.domain.SceneAction;
import edu.ksu.swe4743.smarthomescenes.domain.SceneTarget;
import edu.ksu.swe4743.smarthomescenes.domain.SpecificDeviceTarget;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public final class DeviceSceneBuilder implements SceneBuilder {
    private String name;
    private final List<SceneAction> actions = new ArrayList<>();

    @Override
    public void reset() {
        name = null;
        actions.clear();
    }

    @Override
    public void setName(String name) {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("Scene name is required.");
        }

        this.name = name.trim();
    }

    @Override
    public void addAction(SceneAction action) {
        actions.add(action);
    }

    public DeviceSceneBuilder named(String name) {
        setName(name);
        return this;
    }

    public DeviceSceneBuilder addDeviceAction(
        UUID deviceId,
        String operation,
        Map<String, String> parameters
    ) {
        return addActionInternal(new SpecificDeviceTarget(deviceId), operation, parameters);
    }

    public DeviceSceneBuilder addGroupAction(
        String deviceType,
        String location,
        String operation,
        Map<String, String> parameters
    ) {
        if (deviceType == null || deviceType.isBlank()) {
            throw new IllegalArgumentException("Device type is required.");
        }

        return addActionInternal(new DeviceGroupTarget(deviceType, location), operation, parameters);
    }

    @Override
    public DeviceScene build() {
        if (name == null || name.isBlank()) {
            throw new IllegalStateException("Cannot build a DeviceScene without a name.");
        }

        if (actions.isEmpty()) {
            throw new IllegalStateException("Cannot build a DeviceScene without at least one action.");
        }

        var copiedActions = actions.stream()
            .map(action -> new SceneAction(
                action.getId(),
                action.getOrderIndex(),
                action.getTarget(),
                action.getOperation(),
                new LinkedHashMap<>(action.getParameters())))
            .toList();

        return new DeviceScene(UUID.randomUUID(), name, copiedActions);
    }

    private DeviceSceneBuilder addActionInternal(
        SceneTarget target,
        String operation,
        Map<String, String> parameters
    ) {
        if (operation == null || operation.isBlank()) {
            throw new IllegalArgumentException("Operation is required.");
        }

        var safeParameters = parameters == null
            ? Map.<String, String>of()
            : Map.copyOf(new LinkedHashMap<>(parameters));

        actions.add(new SceneAction(
            UUID.randomUUID(),
            actions.size(),
            target,
            operation.trim(),
            safeParameters));

        return this;
    }
}

File: presentations/14-repository-and-builder-pattern-demos/csharp-smart-home-scenes/Builder/DeviceSceneBuilder.cs
using SmartHomeScenesDemo.Domain;

namespace SmartHomeScenesDemo.Builder;

public sealed class DeviceSceneBuilder : ISceneBuilder
{
    private string? _name;
    private readonly List<SceneAction> _actions = new();

    public void Reset()
    {
        _name = null;
        _actions.Clear();
    }

    public void SetName(string name)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            throw new ArgumentException("Scene name is required.", nameof(name));
        }

        _name = name.Trim();
    }

    public void AddAction(SceneAction action)
    {
        _actions.Add(action);
    }

    public DeviceSceneBuilder Named(string name)
    {
        SetName(name);
        return this;
    }

    public DeviceSceneBuilder AddDeviceAction(
        Guid deviceId,
        string operation,
        IReadOnlyDictionary<string, string>? parameters = null)
    {
        return AddActionInternal(
            new SpecificDeviceTarget(deviceId),
            operation,
            parameters);
    }

    public DeviceSceneBuilder AddGroupAction(
        string deviceType,
        string? location,
        string operation,
        IReadOnlyDictionary<string, string>? parameters = null)
    {
        if (string.IsNullOrWhiteSpace(deviceType))
        {
            throw new ArgumentException("Device type is required.", nameof(deviceType));
        }

        return AddActionInternal(
            new DeviceGroupTarget(deviceType.Trim(), string.IsNullOrWhiteSpace(location) ? null : location.Trim()),
            operation,
            parameters);
    }

    public DeviceScene Build()
    {
        if (string.IsNullOrWhiteSpace(_name))
        {
            throw new InvalidOperationException("Cannot build a DeviceScene without a name.");
        }

        if (_actions.Count == 0)
        {
            throw new InvalidOperationException("Cannot build a DeviceScene without at least one action.");
        }

        var scene = new DeviceScene(
            Guid.NewGuid(),
            _name,
            _actions
                .OrderBy(action => action.OrderIndex)
                .Select(action => new SceneAction(
                    action.Id,
                    action.OrderIndex,
                    action.Target,
                    action.Operation,
                    new Dictionary<string, string>(action.Parameters)))
                .ToArray());

        return scene;
    }

    private DeviceSceneBuilder AddActionInternal(
        SceneTarget target,
        string operation,
        IReadOnlyDictionary<string, string>? parameters)
    {
        if (string.IsNullOrWhiteSpace(operation))
        {
            throw new ArgumentException("Operation is required.", nameof(operation));
        }

        var orderIndex = _actions.Count;
        var safeParameters = parameters is null
            ? new Dictionary<string, string>()
            : new Dictionary<string, string>(parameters);

        _actions.Add(new SceneAction(
            Guid.NewGuid(),
            orderIndex,
            target,
            operation.Trim(),
            safeParameters));

        return this;
    }
}
